src/setfit/onnx.py

```diff
@@ -0,0 +1,227 @@
+from .modeling import SetFitModel
+from sentence_transformers import models
+import numpy as np
+import onnx
+import warnings
+from typing import Optional, Union
+import torch
+from sklearn.linear_model import LogisticRegression
+
+
+class ONNXSetFitModel(torch.nn.Module):
+    def __init__(
+        self,
+        model_body,
+        pooler,
+        model_head: Optional[Union[torch.nn.Module, LogisticRegression]] = None,
+    ):
+        super().__init__()
+
+        self.model_body = model_body
+        self.pooler = pooler
+        self.model_head = model_head
+
+    def forward(self, input_ids, attention_mask, token_type_ids):
+        hidden = self.model_body(input_ids, attention_mask, token_type_ids)
+        hidden = {"token_embeddings": hidden[0], "attention_mask": attention_mask}
+
+        hidden = self.pooler(hidden)["sentence_embedding"]
+        if self.model_head is None:
+            return hidden
+
+        out = self.model_head(hidden)
+        return out
+
+
+def sklearn_head_to_onnx(model_head: LogisticRegression, opset: int):
+    """
+    Convert the sklearn head from a SetFitModel to ONNX format.
+
+    :param model_head: The trained SetFit model_head.
+    :param opset: The ONNX opset to use for optimizing this model. The opset is not
+        gauranteed and will default to the maximum version possible for the sklearn
+        model.
+    :raises ImportError: If skl2onnx is not installed an error will be raised asking
+        to install this package.
+    :return: The ONNX model generated from the sklearn head.
+    """
+
+    # Check if skl2onnx is installed
+    try:
+        from skl2onnx.common.data_types import guess_data_type
+        from skl2onnx import convert_sklearn
+        import onnxconverter_common
+        from skl2onnx.sklapi import CastTransformer
+        from sklearn.pipeline import Pipeline
+    except ImportError:
+        msg = (
+            "skl2onnx must be installed in order to convert a model with an sklearn head."
+        )
+        raise ImportError(msg)
+
+    # Check to see that the head has a coef_
+    if not hasattr(model_head, "coef_"):
+        raise ValueError(
+            "Head must have coef_ attribute check that this is supported by your model and the model has been fit."
+        )
+
+    # Determine the initial type and the shape of the output.
+    input_shape = (None, *model_head.coef_.shape[1:])
+    dtype = guess_data_type(model_head.coef_, shape=input_shape)[0][1]
+    dtype.shape = input_shape
+
+    # If the datatype of the model is double we need to cast the outputs
+    # from the setfit model to doubles for compatibility inside of ONNX.
+    if type(dtype) == onnxconverter_common.data_types.DoubleTensorType:
+        # TODO:: TALK ABOUT FLOAT CONVERSION ISSUES.
```

Comment:
Can you share what this refers to?

Reply:
Revisiting this I don't think it's an issue since we are going float to double. I was worried about doing a type conversion with floats and doubles.  If you used something like a random forest here and did a double to float conversion you could have really weird results since the function isn't continuous.  See this for more details. https://onnx.ai/sklearn-onnx/auto_tutorial/plot_ebegin_float_double.html

I'm going to remove this todo because since we are going from float to double we're increasing precision not decreasing it 😅.